README.md
Windshaft map tiler
===================

[![Build Status](http://travis-ci.org/CartoDB/Windshaft.png)]
(http://travis-ci.org/CartoDB/Windshaft)

A Node.js based webmercator map tile server for PostGIS with Carto map
styling API.

* Pluggable routing to provide customizable tile API URL endpoints
* Before and after filters to allow custom access control and caching
  strategies
* Can render arbitrary SQL queries 
* Generates image and UTFGrid interactivity tiles
* Accepts, stores, serves, and applies map styles written in the Carto
  markup language (same markup as Mapbox Tilemill)
* Accepts custom map styles on a per tile basis in the tile request
* Allows setting of CORS headers to allow access to tile data from
  Javascript

![Puma Concolor by @eightysteele] (https://github.com/CartoDB/Windshaft/raw/master/examples/puma_concolor.png)

Being a dynamic map renderer, windshaft commits some map server 'sins' in
its raw form. The idea is that you the developer will want to graft your
own auth/metrics/caching/scaling on top of decent core components. Same
old story: high cohesion, low coupling makes us happy.

Windshaft is a library used by cartodb.com,
an Open Source Geospatial Database on the Cloud.

Dependencies
------------
* Node >=0.8
* Mapnik 2.0.1, 2.0.2 or 2.1.0 (http://github.com/mapnik/mapnik-reference)
* PostgreSQL >8.3.x, PostGIS >1.5.x
* Redis >2.2.x


Install
-------
```
npm install
```


Usage
-----
```javascript

var Windshaft = require('windshaft');

// Configure pluggable URLs
// =========================
// The config object must define grainstore config (generally just
// postgres connection details), redis config, a base url and a function
// that adds 'dbname' and 'table' variables onto the Express.js req.params
// object.  In this example, the base URL is such that dbname and table will
// automatically be added to the req.params object by express.js. req2params
// can be extended to allow full control over the specifying of database
// parameters and also allows for the req.params object to be extended with
// other variables, such as:
//
// * sql - custom sql query to narrow results shown in map)
// * geom_type - specify the geom type (point|polygon) to get more
//               appropriate default styles
// * cache_buster - forces the creation of a new render object, nullifying
//                  existing metatile caches
// * interactivity - specify the column to use in the UTFGrid
//                   interactivity layer (defaults to null)
// * style - specify map style in the Carto map language on a per tile basis
//
// * dbuser - username for database connection
// * dbpassword - password for database connection
// * dbhost - database host
// * dbport - database port
// * dbname - database name
//
// the base url is also used for persisiting and retrieving map styles via:
//
// GET  base_url + '/style' (returns a map style)
// POST base_url + '/style' (allows specifying of a style in Carto markup
//                           in the 'style' form variable).
//
// beforeTileRender and afterTileRender could be defined if you want yo
// implement your own tile cache policy. See an example below

var config = {
        base_url: '/database/:dbname/table/:table',
        base_url_notable: '/database/:dbname',
        req2params: function(req, callback){
          callback(null,req)
        },
        grainstore: {
          datasource: {
            user:'postgres', host: '127.0.0.1',
            port: 5432
          }
        }, //see grainstore npm for other options
        renderCache: {
          ttl: 60000, // seconds
        },
        mapnik: {
          metatile: 4,
          bufferSize:64
        },
        redis: {host: '127.0.0.1', port: 6379},
        // this two filters are optional
        beforeTileRender: function(req, res, callback) {
            callback(null);
        },
        afterTileRender: function(req, res, tile, headers, callback) {
            callback(null, tile, headers);
        }
    };

// Initialize tile server on port 4000
var ws = new Windshaft.Server(config);
ws.listen(4000);
console.log("map tiles are now being served out of: http://localhost:4000"
            + config.base_url + '/:z/:x/:y.*');

// Specify .png, .png8 or .grid.json tiles.
```

See examples directory for running server and maptile viewer


Installing Mapnik
-----------------

**Source**
https://github.com/mapnik/mapnik/downloads

**OSX**

http://trac.mapnik.org/wiki/MacInstallation/Homebrew

**Linux**

There should be .deb packages out there with stable 2.0.0 packages right now.
Search fro libmapnik2-dev

Tests
-----

Windshaft has a unit and acceptance test suite.
To execute them, run ```make check``` (or ```npm test```).
It is also possible for tests to recurse into some of
the submodules, that's done with ```make check-submodules```
(or ```make check-full``` to run both)

You'll need to be sure your PGUSER (or your libpq default) is
set to a "superuser" PostgreSQL account, for example:

 PGUSER=postgres make check

Troubleshooting
---------------

### Postgres errors persist after configuration change

If you're seeing Postgres errors that you can't dispel through changes in your
Windshaft config (e.g. `ERROR: column "the_geom_webmercator" does not exist`),
try `redis-cli flushall`. Windshaft caches MML files in Redis than can contain SQL
and may not expire them when you change your Windshaft config, but manually flushing 
redis should do the trick.

### Uncaught Error: Command failed running tests
You need [ImageMagick](http://www.imagemagick.org/) to run some tests. In Mac OS X
there are some issues running with latest versions of ImageMagick. If you use
[Homebrew](http://brew.sh/) you can try with a
[modified Formula to install ImageMagick version 6.7.7](https://gist.github.com/rochoa/10017167).

--
Thanks to the Mapnik and Mapbox team for making such flexible tools




NEWS
Version 0.27.1
2014-mm-dd

New features:
 - Add support for raster layers (#190)


Version 0.27.0
2014-09-24

New features:
 - Starts using mapnik 2.3.x with node-mapnik 1.4.15

Enhancements:
 - Upgrades grainstore dependency to be able to use 2.3.0 carto styles


Version 0.26.0
2014-09-18

Enhancements:
 - Torque query improvements
 - New header for database host serving the request
 - Metrics revamp: removes and adds some metrics
    - Profiler is now always available as a proxy so no need to keep checking
      if it exists.
    - Tracks and reports time for torque queries, using chronograph module
 - Stats client now lives in its own file
 - Upgrade dependencies:
    - grainstore
    - tilelive-mapnik
    - tilelive

Bug fixes:
 - Increments counters and track timers only for supported formats
 - Support specifying column name in MapConfig (#191)


Version 0.25.1
2014-08-19

Bug fixes:
 - Responds with 400 error on invalid json input (#156)

Version 0.25.0
2014-08-18

Enhancements:
 - Removes npm-shrinkwrap.json as it was a temporary measure to fix a problem
   with srs.

New features:
 - Deprecates psql and starts using cartodb-psql module


Version 0.24.1
2014-08-13

Enhancements:
 - Checks mapnik version for grainstore is configured, otherwise it uses
   installed one


Version 0.24.0
2014-08-13

 Enhancements:
 - Upgrades dependencies:
    - underscore
    - grainstore
    - redis-mpool
    - pg
 - Specifies name in the redis pool


Version 0.23.0
2014-08-07

 New features:

 - Upgrade to grainstore 0.19.0 version

Version 0.22.0
2014-07-30

 Enhancements:

 - dbhost and dbport are no longer hashed in on the MapConfig identifier hash,
   so dbhost and dbport can be dynamically altered at runtime keeping the same
   URL schema.
 - Cache buster does not invalidate renderer if it is numeric and smaller
   than cached buster (#177)
 - Literal expression for some regular expressions
 - Profiler header sent as JSON string

 Bug fixes:

 - Returns 400 errors for bad filters and missing columns in queries

 Other changes:

 - Metrics documentation

Version 0.21.0
2014-05-07

 Enhancements:

 - Make tests runnable in Mac OS X
 - Code reorganization for CacheEntry
 - assert helper now support per mil tolerance to check image equality

 Bug fixes:

 - Do not omit points close to the tile boundary from torque tiles (#186)
 - Use correct torque.js dependency

Version 0.20.0
2014-03-20

 Enhancements:

 - Do not cache bogus renderers (#171)
 - Upgrade carto to 0.9.5-cdb3
 - Use INFO level for express logging
 - Do not include connection info in error responses (#173)

 Other changes:

 - Use external module as profiler (#174)
 - Switch to 3-clause BSD license (#175)

Version 0.19.4
2014-03-05

 Enhancements:

 - Catch statsd exceptions (#166, #167)

Version 0.19.3
2014-03-04

 Enhancements:

 - Log full connection string on PostgreSQL connection error (#165)
 - Prefix statsd resulting from hits of / and /version endpoints
 - Use 403 status, not 401, for forbidden access
 - Upgrade grainstore to 0.18.1

Version 0.19.2
2014-02-27

 Enhancements:

 - Send 404 instead of 400 on attempts to fetch non-existing database items
 - Send 500 status and better error message on db connection error
 - Use "TorqueRenderer" prefix for errors generated by it (#164)
 - Inject db parameters in user-provided MapConfig objects (#163)

Version 0.19.1
2014-02-26

 Enhancements:

 - Do not send 0 timings to statsd (fixes "Bad line: 0,ms" messages)
 - Do not call req2params more then once on MapConfig creation (#157) 
 - Do not invalidate renderer caches when NO cache_buster is given (#158)
 - Call afterLayergroupCreate only after MapConfig verification (#159)
 - Drop LRU cache for "seen" layergroups (#160)
 - Clearer error on mapnik tokens use with attribute service (#154)
 - Include MapConfig specification and MultilayerAPI documents in repo

 Bug fixes:

 - Fix MapStore.save always false "known" return variable (#162)

Version 0.19.0
2014-02-19

 New features:

 - Use own .sendResponse function for all responses
 - Make MapStore instance accessible via app
 - Allow passing RedisPooler via server options (#146)

 Enhancements:

 - Use log4js express logger if global.log4js is defined (#140)
 - Include format in rendering profile labels (#152)

Version 0.18.2
2014-02-17

 Bug fixes:

 - Fix error message on unexistent map config token (#148)

 Enhancements:

 - Integrate statsD support in Profiler

Version 0.18.1
2014-02-13

 Bug fixes:

 - Fix duplicated garbage collection via grainstore upgrade

 Enhancements:

 - Speed up construction of mapnik renderers
 - Allow disabling statsD via configuration (#144)

Version 0.18.0
2014-02-12

 Enhancements:

 - Advertise support for node-0.10 (#128)
 - Use a single pooler for mapnik and torque renderers (#142)
 - Improve garbage collection performance (#141)
 - Add statsd support

Version 0.17.2
2014-02-11

 Enhancements:

 - Reduce debug logs (#139)

Version 0.17.1
2014-02-11

 Bug fixes:

 - Check attribute service validity at MapConfig creation time (#131)
 - Use read-only transactions for torque tiles and attributes (#130)
 - Fix reading database password (dbpassword) from req.param (#134)

 Enhancements:

 - Add 'torque.json' and 'torque.bin' format aliases (#133)
 - Allow specifying name of torque geometry column in the MapConfig (#125)
 - Improve error message on attempts to fetch torque tile from
   non-torque layers (#136)
 - Include X-Tiler-Profiler header in POST /style endpoint (#138)
 - Improve speed of POST /style and DELETE /style (#138)

Version 0.17.0
2014-02-04

 New features:

 - Configurable endpoint for multilayer API (#126)

 Bug fixes:

 - Fix coordinate order in TorqueRenderer.getTile 

Version 0.16.0
2014-02-04

 New features:

 - Support for torque tiles (#113)
 - New attributes service (#121)

 Enhancements:

 - Store full layergroup configuration in redis (#114)

 Backward incompatible changes:

 - Drop support for XML processor hook (#119)


app.js
var Windshaft = require('./lib/windshaft');
var cluster = require('cluster');
var numCPUs = require('os').cpus().length;
var _          = require('underscore');

// Configure pluggable URLs
// =========================
// The config object must define grainstore config (generally just
// postgres connection details), redis config, a base url and a function
// that adds 'dbname' and 'table' variables onto the Express.js req.params
// object.  In this example, the base URL is such that dbname and table will
// automatically be added to the req.params object by express.js. req2params
// can be extended to allow full control over the specifying of database
// parameters and also allows for the req.params object to be extended with
// other variables, such as:
//
// * sql - custom sql query to narrow results shown in map)
// * geom_type - specify the geom type (point|polygon) to get more
//               appropriate default styles
// * cache_buster - forces the creation of a new render object, nullifying
//                  existing metatile caches
// * interactivity - specify the column to use in the UTFGrid
//                   interactivity layer (defaults to null)
// * style - specify map style in the Carto map language on a per tile basis
//
// * dbuser - username for database connection
// * dbpassword - password for database connection
// * dbhost - database host
// * dbport - database port
// * dbname - database name
//
// the base url is also used for persisiting and retrieving map styles via:
//
// GET  base_url + '/style' (returns a map style)
// POST base_url + '/style' (allows specifying of a style in Carto markup
//                           in the 'style' form variable).
//
// beforeTileRender and afterTileRender could be defined if you want yo
// implement your own tile cache policy. See an example below

var config = {
    base_url: '/database/:dbname/table/:table',
    base_url_notable: '/database/:dbname',
    req2params: function(req, callback){
        // Whitelist query parameters and attach format
        var good_query = ['sql', 'geom_type', 'cache_buster', 'cache_policy', 'callback', 'interactivity', 'map_key', 'api_key', 'auth_token', 'style', 'style_version', 'style_convert', 'config' ];
        var bad_query  = _.difference(_.keys(req.query), good_query);

        _.each(bad_query, function(key){ delete req.query[key]; });
        req.params =  _.extend({}, req.params); // shuffle things as request is a strange array/object
        // bring all query values onto req.params object
        _.extend(req.params, req.query);
        callback(null,req)
    },
    grainstore: {
        datasource: {
            user:'yoan', host: '127.0.0.1',
            port: 5432,
            geometry_field: 'the_geom_webmercator',
            srid: 4326,
            extent: "-20037508.3,-20037508.3,20037508.3,20037508.3",
            //row_limit: 65535,
            //simplify_geometries: true,
            /*
             * Set persist_connection to false if you want
             * database connections to be closed on renderer
             * expiration (1 minute after last use).
             * Setting to true (the default) would never
             * close any connection for the server's lifetime
             */
            persist_connection: false,
            max_size: 500
        }
    }, //see grainstore npm for other options
    renderCache: {
        ttl: 86400000, // milliseconds
    },
    mapnik: {
        metatile: 1,
        bufferSize:32
    },
    redis: {host: '127.0.0.1', port: 6379},
    // this two filters are optional
    beforeTileRender: function(req, res, callback) {
        callback(null);
    },
    afterTileRender: function(req, res, tile, headers, callback) {
        callback(null, tile, headers);
    },
    enable_cors: true
};

if (cluster.isMaster) {
    // Fork workers.
    for (var i = 0; i < numCPUs; i++) {
        cluster.fork();
    }

    cluster.on('exit', function(worker, code, signal) {
        console.log('worker ' + worker.process.pid + ' died');
        cluster.fork();
    });
} else {
    var ws = new Windshaft.Server(config);
    ws.listen(4000);

    // Initialize tile server on port 4000

    console.log("map tiles are now being served out of: http://localhost:4000"
        + config.base_url + '/:z/:x/:y.*');

    // Specify .png, .png8 or .grid.json tiles.

}
